barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/goblin_ultra_circuit_builder.cpp

```diff
@@ -0,0 +1,165 @@
+#include "goblin_ultra_circuit_builder.hpp"
+#include <barretenberg/plonk/proof_system/constants.hpp>
+#include <unordered_map>
+#include <unordered_set>
+
+using namespace barretenberg;
+
+namespace proof_system {
+
+template <typename FF> void GoblinUltraCircuitBuilder_<FF>::finalize_circuit()
+{
+    UltraCircuitBuilder_<FF>::finalize_circuit();
+
+    // Set internally the current and previous size of the aggregate op queue transcript
+    op_queue->set_size_data();
+}
+
+/**
+ * @brief Ensure all polynomials have at least one non-zero coefficient to avoid commiting to the zero-polynomial
+ *
+ * @param in Structure containing variables and witness selectors
+ */
+// TODO(#423): This function adds valid (but arbitrary) gates to ensure that the circuit which includes
+// them will not result in any zero-polynomials. It also ensures that the first coefficient of the wire
+// polynomials is zero, which is required for them to be shiftable.
+// TODO(luke): Add ECC op gate to ensure op wires are non-zero?
```

Comment:
Those would accumulate( Technically, we should never use GolbinUltra if we are not feeding a proof inside anyway, right?

Reply:
Yeah this is an old TODO. deleted